p4_symbolic/ir/ir.cc

```diff
@@ -196,6 +425,56 @@ gutil::StatusOr<RValue> ExtractRValue(
       variable->set_name(variables[variable_index]);
       break;
     }
+    case bmv2::ExpressionType::hexstr_: {
+      HexstrValue *hexstr_value = output.mutable_hexstr_value();
+      std::string hexstr = struct_value.fields().at("value").string_value();
+      if (absl::StartsWith(hexstr, "-")) {
+        hexstr_value->set_value(std::string(absl::StripPrefix(hexstr, "-")));
+        hexstr_value->set_negative(true);
+      } else {
+        hexstr_value->set_value(hexstr);
+        hexstr_value->set_negative(false);
+      }
+      break;
+    }
+    case bmv2::ExpressionType::bool_: {
+      output.mutable_bool_value()->set_value(
+          struct_value.fields().at("value").bool_value());
+      break;
+    }
+    case bmv2::ExpressionType::string_: {
+      output.mutable_string_value()->set_value(
+          struct_value.fields().at("value").string_value());
+      break;
+    }
+    case bmv2::ExpressionType::expression: {
+      const google::protobuf::Struct *expression =
```

Comment:
I wanted to do that initially, however that won't work.

The problem here is the while loop that follows. The loop keeps on recursively extracting nested expression from inside this expression, until the base case is reached. This is needed because p4c sometimes produces weirdly nested expressions on the form
{
 "type": "expression",
 "value": {
     "type": "expression",
     "value": {
       .....
           {
            "op": "+",
            "left": <x>,
            "right": <y>
           }
       ....
    }
}

I think of this as the bmv2 format equivalent of something like (((x+y)))

Regardless, the point here is that this expression is not necessarily the expression that we need, and we will need to re-assign it to its child expression in that case, and repeated, until we find the base case. I can't re-assign references that way, I do not think copying this over and over again would be nice either.